### PR TITLE
Resilience4j bulkhead patterns

### PR DESCRIPTION
**Pull request by Ferioney**
added supporting of resilience4j bulkhead patterns
Fixes gh-86

### CODE REVIEW

**Review by ryanjbaxter — CHANGES_REQUESTED** [review]
We cannot make such major changes in a minor release.  We need to maintain the default behavior.

I think we can optionally enable the use of a Bulkhead if `resilience4j-bulkhead` is on the classpath.

File: spring-cloud-circuitbreaker-resilience4j/pom.xml [hunk 1]
```diff
@@ -27,6 +27,11 @@
 			<groupId>io.github.resilience4j</groupId>
 			<artifactId>resilience4j-timelimiter</artifactId>
 		</dependency>
+		<dependency>
+			<groupId>io.github.resilience4j</groupId>
+			<artifactId>resilience4j-all</artifactId>
+			<scope>provided</scope>
```
Comment (ryanjbaxter) [inline spring-cloud-circuitbreaker-resilience4j/pom.xml:33]: shouldn't this be `optional`?

File: spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreaker.java [hunk 2]
```diff
@@ -48,17 +52,18 @@
 
 	private Optional<Customizer<io.github.resilience4j.circuitbreaker.CircuitBreaker>> circuitBreakerCustomizer;
 
-	public Resilience4JCircuitBreaker(String id,
-			io.github.resilience4j.circuitbreaker.CircuitBreakerConfig circuitBreakerConfig,
+	public Resilience4JCircuitBreaker(String id, CircuitBreakerConfig circuitBreakerConfig,
```
Comment (ryanjbaxter) [inline spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreaker.java:55]: Add a separate constructor

File: spring-cloud-circuitbreaker-resilience4j/pom.xml (hunk 1 above)
Comment (ryanjbaxter) [reply spring-cloud-circuitbreaker-resilience4j/pom.xml:33]: Why not use `resilience4j-bulkhead`?

**Review by ryanjbaxter — CHANGES_REQUESTED** [review]

File: spring-cloud-circuitbreaker-resilience4j/pom.xml (hunk 1 above)
Comment (Ferioney) [reply spring-cloud-circuitbreaker-resilience4j/pom.xml]: yep, changed.

`resilience4j-all` provides a Decorator class which is very useful for creating citcuitBreater/TimeLimiter/Bulkheads pipeline. 
And `resilience4j-all` includes `resilience4j-bulkhead` module

Comment (ryanjbaxter) [reply spring-cloud-circuitbreaker-resilience4j/pom.xml]: Do you know which module the Decorator is in?  `-all` brings in much more than we need.

File: spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreaker.java (hunk 2 above)
Comment (ryanjbaxter) [reply spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreaker.java:68]: Bump

File: spring-cloud-circuitbreaker-resilience4j/pom.xml (hunk 1 above)
Comment (Ferioney) [reply spring-cloud-circuitbreaker-resilience4j/pom.xml]: In `resilience4j-all`: [Decorators](https://github.com/resilience4j/resilience4j/blob/master/resilience4j-all/src/main/java/io/github/resilience4j/decorators/Decorators.java)

Comment (ryanjbaxter) [reply spring-cloud-circuitbreaker-resilience4j/pom.xml]: @Ferioney how much extra code would it be to not use the decorators?  I talked this over with the team and the general consensus was that we should avoid requiring extra dependencies when they are not needed.

Comment (Ferioney) [reply spring-cloud-circuitbreaker-resilience4j/pom.xml]: I think we can do it without Decorators:
[Commit 870dc37 Resilience4jBulkheadProvider.java](https://github.com/spring-cloud/spring-cloud-circuitbreaker/pull/87/commits/870dc37bd1bacb0f6d52db5d00e4a79063b43b51#diff-6ddba7b121b4caa51fd8c5b41821a64ce0d7d52458b3ace890911c51d41d06feR108)

**Comment by Ferioney** [comment]
@ryanjbaxter  I think we can also implement support of config files in this topic (similar to the [Support Resilience4J Configuration Properties](https://github.com/spring-cloud/spring-cloud-circuitbreaker/pull/88) ). What do you think?

**Review by ryanjbaxter — APPROVED** [review]

**Comment by ryanjbaxter** [comment]
@Ferioney This looks good just fix the merge conflicts.

> I think we can also implement support of config files in this topic (similar to the Support Resilience4J Configuration Properties ). What do you think?

So you can configure bulkheads using configuration properties?

**Comment by Ferioney** [comment]
@ryanjbaxter I did some major changes in PR, please review again.
Main changes:
1. Added support of properties file for Bulkheads.
2. Changed default bulkhead implementation from Semaphone to ThreadPool. My team and I think it more relevant for people who are going to migrate from Hystrix to Resilience4j.

File: README.adoc [hunk 3]
```diff
@@ -138,6 +138,105 @@ public Customizer<ReactiveResilience4JCircuitBreakerFactory> slowCustomizer() {
 ----
 ====
 
+==== Bulkhead pattern supporting
+If `resilience4j-bulkhead` is on the classpath, Spring Cloud Circuit Breaker will wrap all methods with a Resilience4j Bulkhead.
```
Comment (ryanjbaxter) [inline README.adoc:142]: Change "Circuit Breaker" to "CircuitBreaker" in the docs

File: spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JAutoConfiguration.java [hunk 4]
```diff
@@ -52,13 +57,33 @@
...
 		Resilience4JCircuitBreakerFactory factory = new Resilience4JCircuitBreakerFactory(circuitBreakerRegistry,
-				timeLimiterRegistry);
+				timeLimiterRegistry, bulkheadProvider);
 		customizers.forEach(customizer -> customizer.customize(factory));
 		return factory;
 	}
 
+	@Configuration(proxyBeanMethods = false)
+	@ConditionalOnClass(Bulkhead.class)
+	@ConditionalOnProperty(value = "spring.cloud.bulkhead.resilience4j.enable", matchIfMissing = true)
```
Comment (ryanjbaxter) [inline spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JAutoConfiguration.java:70]: I think the property should be `spring.cloud.circuitbreaker.bulkhead.resilience4j.enable`

It also needs to be added to this file

https://github.com/spring-cloud/spring-cloud-circuitbreaker/blob/master/spring-cloud-circuitbreaker-resilience4j/src/main/resources/META-INF/additional-spring-configuration-metadata.json

**Review by ryanjbaxter — CHANGES_REQUESTED** [review]

File: spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JAutoConfiguration.java (hunk 4 above)
Comment (spencergibb) [reply spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JAutoConfiguration.java]: enabled is the convention

**Review by ryanjbaxter — APPROVED** [review]

File: spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JAutoConfiguration.java (hunk 4 above)
Comment (ryanjbaxter) [reply spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JAutoConfiguration.java]: Good call!

File: spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JAutoConfiguration.java [hunk 5]
```diff
@@ -52,13 +57,33 @@
...
 		Resilience4JCircuitBreakerFactory factory = new Resilience4JCircuitBreakerFactory(circuitBreakerRegistry,
-				timeLimiterRegistry);
+				timeLimiterRegistry, bulkheadProvider);
 		customizers.forEach(customizer -> customizer.customize(factory));
 		return factory;
 	}
 
+	@Configuration(proxyBeanMethods = false)
+	@ConditionalOnClass(Bulkhead.class)
+	@ConditionalOnProperty(value = "spring.cloud.circuitbreaker.bulkhead.resilience4j.enable", matchIfMissing = true)
```
Comment (ryanjbaxter) [inline spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JAutoConfiguration.java:70]: Change to `.enabled`

**Pull request merged**